src/raft.rs

```diff
@@ -2096,176 +2151,154 @@ impl<T: Storage> Raft<T> {
-            self.mut_prs().promote_learner(id)
-        } else {
-            let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
-            self.mut_prs().insert_voter(id, progress)
-        };
-
-        if let Err(e) = result {
-            error!(self.logger, ""; "e" => %e);
-            return Err(e);
-        }
-        if self.id == id {
-            self.promotable = !learner;
-        };
-        // When a node is first added/promoted, we should mark it as recently active.
-        // Otherwise, check_quorum may cause us to step down if it is invoked
-        // before the added node has a chance to communicate with us.
-        self.mut_prs().get_mut(id).unwrap().recent_active = true;
-        result
-    }
-
-    /// Adds a new node to the cluster.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is already a voter.
-    /// * `id` is already a learner.
-    pub fn add_node(&mut self, id: u64) -> Result<()> {
-        self.add_voter_or_learner(id, false)
-    }
-
-    /// Adds a learner node.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is already a voter.
-    /// * `id` is already a learner.
-    pub fn add_learner(&mut self, id: u64) -> Result<()> {
-        self.add_voter_or_learner(id, true)
+        true
     }
 
-    /// Removes a node from the raft.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is not a voter or learner.
-    pub fn remove_node(&mut self, id: u64) -> Result<()> {
-        self.mut_prs().remove(id)?;
-
-        // do not try to commit or abort transferring if there are no voters in the cluster.
-        if self.prs().voter_ids().is_empty() {
-            return Ok(());
+    /// Updates the in-memory state and, when necessary, carries out additional actions
+    /// such as reacting to the removal of nodes or changed quorum requirements.
+    pub fn post_conf_change(&mut self) -> ConfState {
+        info!(self.logger, "switched to configuration"; "config" => ?self.prs.conf());
+        // TODO: instead of creating a conf state, validating conf state inside
+        // progress tracker is better.
+        let cs = self.prs.conf().to_conf_state();
+        let is_voter = self.prs.conf().voters.contains(self.id);
+        self.promotable = is_voter;
+        if !is_voter && self.state == StateRole::Leader {
+            // This node is leader and was removed or demoted. We prevent demotions
+            // at the time writing but hypothetically we handle them the same way as
+            // removing the leader: stepping down into the next Term.
+            //
+            // TODO(tbg): step down (for sanity) and ask follower with largest Match
+            // to TimeoutNow (to avoid interruption). This might still drop some
+            // proposals but it's better than nothing.
+            //
+            // TODO(tbg): test this branch. It is untested at the time of writing.
+            return cs;
+        }
```

Comment:
Can do some "safe operations" to evict leader itself e.g. `r.become_follower` first?  How does TiKV handle such a situation currently or maybe PD can just avoid this?

Reply:
TiKV rejects removing leader. It simply drops any proposal that tries to remove itself. However, it's not guaranteed.